### DISCUSSION TITLE
Type check for variable assignment

### DISCUSSION THREAD

**Discussion by k4ml** [beartype/beartype#231] [General]
This will give errors when running with mypy:-

```
from beartype import beartype

class TP: pass

@beartype
def add(x: int, y: int) -> TP:
    return TP()

@beartype
def main():
    total: str = add(1, 2)

main()
```

```
poetry run mypy t.py 
t.py:11: error: Incompatible types in assignment (expression has type "TP", variable has type "str")
```

But beartype doesn't throw any error when actually running it.

**Comment by leycec** [comment]
Thanks so much for the awesome question, @k4ml. You're *absolutely* right, of course.

@beartype currently lacks support for [PEP 526 (AKA, variable annotations)](https://peps.python.org/pep-0526). We're tracking @beartype support for PEP 526 at [long-standing feature request #105](https://github.com/beartype/beartype/issues/105). I've been debating whether to tackle #105 now or later. This seems to be *much* more popular than I'd thought. I'm edging towards: "Now! Gods! Oh, Gods above! Do this!"

Until I do something, there *is* a temporary solution. I almost regret mentioning this, because this requires manual boilerplate. But... @beartype's [shiny DOOR API *does* provide an out-of-the-box solution for manually type-checking anything at any time: the `beartype.door.die_if_unbearable()` function](https://beartype.readthedocs.io/en/latest/api_door/#beartype.door.die_if_unbearable). By manually calling that function after every variable annotation of interest, you can force @beartype to type-check variable assignments:

```python
from beartype import beartype
from beartype.door import die_if_unbearable

class TP: pass

@beartype
def add(x: int, y: int) -> TP:
    return TP()

@beartype
def main():
    total: str = add(1, 2)
    die_if_unbearable(total, str)  # <-- this is the magic. but... it's not so magical.

main()
```

@beartype now raises the expected exception:

```
beartype.roar.BeartypeDoorHintViolation: Object <__main__.TP object at 0x7f08f55d6310>
violates type hint <class 'str'>, as <class "__main__.TP"> <__main__.TP object at
0x7f08f55d6310> not instance of str.
```

I know. *I know.* Having to manually call `die_if_unbearable()` everywhere sucks. Nobody wants to do something that sucks. Let's see if I can finally finish the `beartype.claw` API and get this working for everybody... `</sigh>`

**Reply by k4ml** [reply]
Aaah, so that's how it being used. I read that in the docs and somehow have a feeling this could be what I'm looking for but. I thought simply importing die_if_unbearable will trigger the necessary machinery for the type check.

**Reply by leycec** [reply]
**Yup!** You got it. Really sorry for the confusion there, @k4ml. Our documentation could clearly use some work – lots and lots of back-breaking work that will help me build character, a muscular body, and inner peace. :sweat_smile: 

I should definitely improve our documentation for `die_if_unbearable()` with a similar example. If you have this question, everybody else has this question too. Let's see if I do anything other than read endless volumes of manga tonight. The struggle is real, folks.
